core/primitives/src/types.rs

```diff
@@ -43,6 +42,45 @@ impl<'a> From<&'a ReadablePublicKey> for PublicKey {
 pub type ReceiptId = Vec<u8>;
 pub type CallbackId = Vec<u8>;
 
+#[derive(Clone, Debug, Serialize, Deserialize, PartialEq, Eq)]
+pub struct GroupSignature {
+    #[serde(with = "bs58_serializer")]
+    pub signature: BlsSignature,
+    pub authority_mask: AuthorityMask,
+}
+
+impl GroupSignature {
+    // TODO (optimization): It's better to keep the signature in projective coordinates while
+    // building it, then switch to affine coordinates at the end.  For the time being we just keep
+    // it in affine coordinates always.
+    pub fn add_signature(&mut self, signature: &PartialSignature, authority_id: usize) {
```

Comment:
This usize catches an eye here. I see that you use it from .enumerate below, but may be just use u32 here and `as u32` there?